samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -604,6 +597,9 @@ private StorageEngine createStore(String storeName, TaskName taskName, TaskModel
           Map<String, StorageEngine> sideInputStores = getSideInputStores(taskName);
           Map<String, Set<SystemStreamPartition>> sideInputStoresToSSPs = new HashMap<>();
 
+          CountDownLatch taskCountDownLatch = new CountDownLatch(1);
+          this.sideInputTaskLatches.put(taskName, taskCountDownLatch);
```

Comment:
IMO this is safer as it eliminates sharing of objects between class instances. E.g. in the shared case a task could mistakenly `countDown` the latch more than once, which would be logically incorrect. The way it's written now, that would not be possible.

I think it is also more readable; a dev can read through and understand the latch within `TaskSideInputHandler` without needing to know the relationship to CSM.

I don't see an obvious shortcoming to the approach I've taken, can you explain if your preference is to use one latch and why?

Reply:
Can you elaborate on this if this is a blocking issue?

Reply:
Now that I think about it a bit more, I think you can do away without latches here. Since you will anyways need some form of callback to the runloop within CSM to denote completion of a message, can you actually not piggy back on the callback for particular end offset to count down the latches? With this, you can reuse the same approach for transactional state support for standbys.

Unless you have a rationale of handling that part differently and not tie that into the initial catch up scenario.

Reply:
did we resolve this offline? i remembering discussing that callbacks wouldn't quite work. ok to leave as-is?

Reply:
yeah. we can leave it as is.